crates/ruff_db/src/vendored.rs

```diff
@@ -0,0 +1,379 @@
+            || archive
+                .lookup_path(&normalized.with_trailing_slash())
+                .is_ok()
+    }
+
+    /// Return the crc32 hash of the file
+    pub fn file_hash(&self, path: &VendoredPath) -> Option<u32> {
+        let normalized = normalize_vendored_path(path);
+        let inner_locked = self.inner.lock();
+        let mut archive = inner_locked.borrow_mut();
+        if let Ok(zip_file) = archive.lookup_path(&normalized) {
+            return Some(zip_file.crc32());
+        }
+        archive
+            .lookup_path(&normalized.with_trailing_slash())
+            .map(|zip_file| zip_file.crc32())
+            .ok()
+    }
+
+    /// Read the entire contents of the path into a string
+    pub fn read(&self, path: &VendoredPath) -> Option<String> {
+        let inner_locked = self.inner.lock();
+        let mut archive = inner_locked.borrow_mut();
+        let mut zip_file = archive.lookup_path(&normalize_vendored_path(path)).ok()?;
+        let mut buffer = String::new();
+        zip_file.read_to_string(&mut buffer).ok()?;
+        Some(buffer)
+    }
+}
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash, is_macro::Is)]
+pub enum FileType {
+    /// The path exists in the zip archive and represents a vendored file
+    File,
+
+    /// The path exists in the zip archive and represents a vendored directory of files
+    Directory,
+}
+
+#[derive(Debug)]
+struct VendoredFileSystemInner(Mutex<RefCell<VendoredZipArchive>>);
+
+impl VendoredFileSystemInner {
+    fn new(raw_bytes: &'static [u8]) -> Result<Self> {
+        Ok(Self(Mutex::new(RefCell::new(VendoredZipArchive::new(
+            raw_bytes,
+        )?))))
+    }
+
+    /// Acquire a lock on the underlying zip archive.
+    /// The call will block until it is able to acquire the lock.
+    ///
+    /// ## Panics:
+    /// If the current thread already holds the lock.
+    fn lock(&self) -> LockedZipArchive {
+        LockedZipArchive(self.0.lock().unwrap())
+    }
+}
+
+#[derive(Debug)]
+struct LockedZipArchive<'a>(MutexGuard<'a, RefCell<VendoredZipArchive>>);
+
+impl<'a> LockedZipArchive<'a> {
+    fn borrow_mut(&self) -> ArchiveReader {
+        ArchiveReader(self.0.borrow_mut())
+    }
+}
+
+#[derive(Debug)]
+struct ArchiveReader<'a>(RefMut<'a, VendoredZipArchive>);
+
+impl<'a> Deref for ArchiveReader<'a> {
+    type Target = VendoredZipArchive;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl<'a> DerefMut for ArchiveReader<'a> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.0
+    }
+}
+
+/// Newtype wrapper around a ZipArchive.
+#[derive(Debug)]
+struct VendoredZipArchive(ZipArchive<io::Cursor<&'static [u8]>>);
+
+impl VendoredZipArchive {
+    pub fn new(data: &'static [u8]) -> Result<Self> {
+        Ok(Self(ZipArchive::new(io::Cursor::new(data))?))
+    }
+
+    fn lookup_path(&mut self, path: &NormalizedVendoredPath) -> Result<ZipFile> {
+        Ok(self.0.by_name(path.as_str())?)
+    }
```

Comment:
Should `lookup_path` take a `&VendoredPath` instead and do all the normalization?

Reply:
Because of the "trailing slashes are semantically significant for zip archives" thing combined with the "`ZipFile::by_name()` requires a mutable reference" thing, it's hard for me to see how we could do this while keeping the borrow checker happy and avoiding doing path normalization twice. It doesn't seem possible to call `self.0.by_name()` twice in `lookup_path()` while keeping the borrow checker happy, unfortunately (unless I'm missing something), and because of the trailing-slash thing, we need to call it twice in `lookup_path()` if we're going to do the normalization inside `lookup_path()`.